### ISSUE TITLE
document that / should be read-only on the initrd transition, or /var already mounted

### ISSUE THREAD

**Issue by steelcowboy** [systemd/systemd#832]
I have /var on a separate partition and a separate disk. In /etc/systemd/journal.conf I set "Storage=persistent" yet when I boot up /var/journal/<string>/system.journal never gets written to and running journalctl as root does not show messages as root.

The fix I used for this was to add `var.mount` to the `After=` section of systemd-journald.service. However, if Storage is set to persistent and the file it wants to write to is not available it should wait for the file to become available before it writes the journal; I'd assume it is falling back to temporary storage instead when this happens. 

I'm not sure if this is a bug or just a misconfiguration on my part, but I do think something's wrong since this race condition is occurring.

**Comment by arvidjaar** [comment]
> The fix I used for this was to add var.mount to the After= section of systemd-journald.service. 
> 
> ```
> RequiresMountsFor=/var/log/journal
> ```
> 
> is probably more appropriate.

**Comment by poettering** [comment]
It's the job of `systemd-journal-flush.service` to tell journald to flush the journal data from /run to /var. It does so by sending SIGUSR1 to it. Does this service work correctly for you? What does `systemctl status systemd-journal-flush.service` output for you?

Note that s-j-f.s is ordered after `local-fs.target`, which is ordered after `/var`. It is not OK to order journald itself after /var, since pretty much all services and tools need to connect to journald to work correctly, including the mount command for /var itself, because stdout/stderr is passed to the journal, to ensure everything is logged.

Please provide your `/etc/fstab`, too. 

If this is a problem, and you send another SIGUSR1 to journald, is the issue fixed then?

**Comment by poettering** [comment]
And, no, `RequiresMountsFor=/var/log/journal` in `systemd-journald.service` is not more appropriate either...

**Comment by steelcowboy** [comment]
Alright, so I just removed the `var.mount` part from the service. I confirmed that Storage=persistent is set. Running `systemctl status systemd-journal-flush.service` shows the message:

```
 ● systemd-journal-flush.service - Flush Journal to Persistent Storage
     Loaded: loaded (/usr/lib/systemd/system/systemd-journal-flush.service; static; vendor preset: disabled)
     Active: active (exited) since Mon 2015-08-03 19:31:11 PDT; 2min 17s ago
     Docs: man:systemd-journald.service(8)
       man:journald.conf(5)
     Process: 605 ExecStart=/usr/bin/journalctl --flush (code=exited, status=0/SUCCESS)
     Main PID: 605 (code=exited, status=0/SUCCESS)
     CGroup: /system.slice/systemd-journal-flush.service

     Warning: Journal has been rotated since unit was started. Log output is incomplete or unavailable.  
```

However, the service does not show that it failed. Running journalctl as root only shows services as my user and no system messages.  

That's odd... when I used var.mount running systemctl status <service> and journalctl -b I would not get any output from system processes unless I ran as root -- same thing before I started having this issue. However, I actually killed the journal process (not sent SIGUSR1) and it restarted and now everything is working just like it did pre-July 19th, when my issue started. Sending it SIGUSR1 does not make it flush to journal.

Here's my fstab: https://steelcowboy.me/fstab.txt

**Comment by arvidjaar** [comment]
Does file /run/systemd/journal/flushed exist before or after SIGUSR1? is there anything under /run/log/journal immediately after start? What `systemctl --no-pager show -p After -p Requires systemd-journal-flush.service` show?

**Comment by steelcowboy** [comment]
1. /run/systemd/journal/flushed exists both before and after SIGUSR1
2. There does not seem to be anything under /run/log/journal
3. Requires=-.mount var.mount systemd-journald.service
   After=systemd-journald.service systemd-journald.socket systemd-remount-fs.service var.mount system.slice -.mount`

**Comment by arvidjaar** [comment]
Are there any files with timestamp newer than boot immediately after reboot in `/var/log/journal/long-long-uuid`?

**Comment by poettering** [comment]
Hmm. Maybe there is just some confusion around where the journal stores its files. They are located under /var/log/journal/$MACHINEID/.

Can you paste the full list of files underneath /var/log/journal and all its subdirs please?

**Comment by poettering** [comment]
Hmmm. also can you check if the files are possibly stored underneat the /var mount point on the real root fs? i figure the journal starts writing there during early boot and never bothers looking at anything else...

Do you have "ro" on the kernel cmdline?

**Comment by steelcowboy** [comment]
Here is everything under /var/log/journal:  

```
steelcowboy@athena ~ $ tree /var/log/journal
/var/log/journal
└── fb0cfc4d61a346a5b9bcff2a6f8eb362
    ├── system@00051c5db562c51a-ddadaf7fdd7441af.journal~
    ├── system.journal
    ├── user-1000@00051c5db62ef795-355c2ff6846f1d8f.journal~
    └── user-1000.journal

1 directory, 4 files
```

Ahhhh. After unmounting /var and looking under the root file system, here's what I have:

```
journal
├── fb0cfc4d61a346a5b9bcff2a6f8eb362
│   ├── system@00051b006c1f1bd3-ed75c8451adc7c7e.journal~
│   ├── system@00051b007d870dc6-74d421296eeff741.journal~
│   ├── system@00051b3e98aada06-1fbf3cc1d9ffdc34.journal~
│   ├── system@00051c1add63d098-b2d3fa54370d40c3.journal~
│   ├── system@00051c48d43d3903-f0184c044a9bf462.journal~
│   ├── system.journal
│   └── user-1000.journal
└── remote

2 directories, 7 files
```

And running journalctl _now_ shows me the boot messages -- even up to when I unmounted. So somehow there exist parallel /var's and journald continues to write to the one on my SSD instead of the one on my HDD, but the user journal writes to the /var on the HDD...

**Comment by arvidjaar** [comment]
But how it can work with separate /var? On startup server immediately creates /var/log/journal and fllushes journal to persistent storage independently of any systemd-journal-flush.service

```
        r = server_init(&server);
        if (r < 0)
                goto finish;

        server_vacuum(&server);
        server_flush_to_var(&server);
```

**Comment by poettering** [comment]
Well, the idea is that /var is read-only until it is properly mounted. Hence I was asking if @steelcowboy has "ro" on the kernel command line. It appears to me that he does not, and hence /var is writable right-away which it really shouldn't be.

Or in other words: either the initrd mounts /var into place, so that journald on the host can just make use of it. Or it does not mount /var into place in which case / should stay "ro" until systemd-remount-fs.service remounts it writable, and /var is possible overmounted with the real fs to use.

Hence, do you specify "ro" on the kernel cmdline? if not, why not? which initrd implementation do you use?

THis really sounds like an unexpected behaviour in the initrd. Admittedly though I doubt this is currently not documented, and we should...

**Comment by steelcowboy** [comment]
What do you mean by which initrd do I use? I'm on Arch and use systemd with a standard systemd install.

On my old computer I had a separate var partition and the journal never had any issues... Though it was all on an HDD and nowhere as fast as this system.

If I use ro in my command line will everything get remounted by fstab as rw? Maybe if you want to make it so it would work without the ro by default you could make the journal flush service ensure that if /var is on a separate partition wait until it gets mounted to flush? That should solve these problem I'd think

**Comment by arvidjaar** [comment]
> initrd mounts /var into place

Where this is stated? Which initrd implementation does it?

> do you specify "ro" on the kernel cmdline?

We had (more than one) long discussion on this and common opinion was that this is archaic and outdated and should not be used today. This is the first time I see as requirement from _systemd_ side.

**Comment by poettering** [comment]
@steelcowboy well, the difference is that in systemd logging is provided from the earliest moment on, while on sysvinit it was a late boot service that was only started after all partitions were mounted. journald starts very early and tries to make use of /var right-away if it can, to avoid the complexity of later flushing /run to /var if possible. The benefit of starting the journal this early is that the mounting of /var and so on itself can be logged properly.

But yes, the idea is that "ro" on the kernel command line means the root fs is mounted read-only initially and then through /etc/fstab is mounted writable when ready. After that we then flush the journal from /run to /var. The problem in your case is that journald right from the start can access /var in writable mode and hence makes direct use of it, so that the whole flushing can be avoided...

@arvidjaar: yes as mentioned this behaviour is not documented, but we should as part of the initrd interface wiki text. Who is "we" in your sentence? upstream or downstream? I think I always said that "ro" is the way to go and leave mounting rw to the host after mounting everything, or mounting everything into place from the fstab in which case "rw" is fine. But mounting some stuff from the initrd and some stuff not, but then also making it writable right-away is not really supported (and we should document that...)

**Comment by mbiebl** [comment]
@poettering are you saying, that if you have an initrd which mounts (and fscks) / and possibly /usr, it should either mount them ro or also mount /var if it's on a separate partition? That is certainly a new requirement and one which will be hard to implement, assuming /var can be remote.

It seems to me, that journald should _not_ start wrinting to /var directly as soon as it finds a writable /var, but only do so, after it was been signaled by systemd-journal-flush.service.

**Comment by poettering** [comment]
@mbiebl yupp, that's what I am saying: "ro" should be the default on the kernel cmdline like it always was, and initrd should honour it, and leave it to systemd-remount-fs to then mount it writable later on.

**Comment by arvidjaar** [comment]
@poettering 

> Who is "we" in your sentence? upstream or downstream?

Upstream. The issue was discussed as part of systemd-remount-fs problem (lack of synchronization between initrd and systemd), and it was considered that mounting ro and remounting rw is dangerous and may not even be supported by modern filesystems so root must be mounted rw by initrd.

> [initrd] ... mounting everything into place from the fstab in which case "rw" is fine

I doubt it is really possible. It may require arbitrary deep storage and networking stack that simply may not be available in initrd.

**Comment by poettering** [comment]
"dangerous"? Where was that discussion? Was I part of this? Have a link?

I am not sure I understand what you mean by "arbitrary deep storage"?

Just mount everything from the initrd ro if "ro" is on the kernel cmdline, and let s-r-f.s remount it writable after the transition...

**Comment by arvidjaar** [comment]
> "dangerous"? Where was that discussion? Was I part of this? Have a link?

I may have mixed up things here but from memory it was about fsck'ing mounted root, even if mounted just read-only; and consensus was that initrd should fsck and mount it. Anyway two of these threads are

http://lists.freedesktop.org/archives/systemd-devel/2015-January/026818.html
http://lists.freedesktop.org/archives/systemd-devel/2014-July/020842.html

And in the former you said "I think on Fedora the fsck isn't invoked twice because the initrd already
mounts / writable" ...

> I am not sure I understand what you mean by "arbitrary deep storage"?

LVM on Linux MD over FCoE ... 

> Just mount everything from the initrd ro if "ro" is on the kernel cmdline

Ignoring rather vague "everything" - and what if user did **not** use "ro" on kernel cmdline? What you suggest translates to - journald always starts with in-memory journal and flushes it to stable storage only later. Then simply remove this micro-optimization from journald startup because your suggestion will make this code never execute anyway.

**Comment by poettering** [comment]
Hmm, if dracut on FEdora always mounts / writable, then the split-out /var is indeed also broken on Fedora...

@haraldh could you comment on this?

**Comment by davispuh** [comment]
This is clearly systemd's bug. When have `/var` on separate filesystem then it will write journal to `/var` before it's mounted.

On Arch Linux with just compiled systemd.

```
$ journalctl -b
Failed to get journal fields: Cannot assign requested address
-- Logs begin at Tue 2015-06-23 05:43:19 EEST, end at Sun 2015-08-30 06:09:20 EEST. --
```

kernel parameters

```
ro root=UUID=75543028-e22c-45e8-9e60-73eb43aeeab5
```

`/var` on root filesystem (it writes even when it's chmod 000 )

```
d--------- 1 root root   6 aug 30 06:10 var/

/var/
└── log
    └── journal
        └── 873a5f55f2aa4b33b2568baca40e6a91
            └── system.journal
```

`/etc/fstab`

```
UUID=75543028-e22c-45e8-9e60-73eb43aeeab5   /      btrfs rw,relatime,autodefrag,compress=lzo      0 0
UUID=247e6249-6de1-45cb-9dd0-fa8a654234bf   /var   btrfs rw,noatime,autodefrag,compress=lzo       0 0
```

Initramfs uses systemd so it should see that fstab contains `/var` and should mount it before switching root because seems that journal flushes to `/var` when switching root.

From that journal on root filesystem

```
systemd[1]: Found device SAMSUNG_SP1213C.
systemd[1]: Starting File System Check on /dev/disk/by-uuid/75543028-e22c-45e8-9e60-73eb43aeeab5...
systemd[1]: Started File System Check on /dev/disk/by-uuid/75543028-e22c-45e8-9e60-73eb43aeeab5.
systemd[1]: Mounting /sysroot...
systemd[1]: Mounted /sysroot.
systemd[1]: Reached target Initrd Root File System.
systemd[1]: Starting Initrd Root File System.
systemd[1]: Starting Reload Configuration from the Real Root...
systemd[1]: Reloading.
systemd[1]: Started Reload Configuration from the Real Root.
systemd[1]: Reached target Initrd File Systems.
systemd[1]: Starting Initrd File Systems.
systemd[1]: Reached target Initrd Default Target.
systemd[1]: Starting Initrd Default Target.
systemd[1]: Starting Cleaning Up and Shutting Down Daemons...
systemd[1]: Stopped target Timers.
systemd[1]: Stopping Timers.
systemd[1]: Stopped Cleaning Up and Shutting Down Daemons.
systemd[1]: Stopped target Initrd Default Target.
systemd[1]: Stopping Initrd Default Target.
systemd[1]: Stopped target Basic System.
systemd[1]: Stopping Basic System.
systemd[1]: Stopped target System Initialization.
systemd[1]: Stopping System Initialization.
systemd[1]: Stopped udev Coldplug all Devices.
systemd[1]: Stopping udev Coldplug all Devices...
systemd[1]: Stopping udev Kernel Device Manager...
systemd[1]: Starting Setup Virtual Console...
systemd[1]: Stopped target Swap.
systemd[1]: Stopping Swap.
systemd[1]: Stopped target Local File Systems.
systemd[1]: Stopping Local File Systems.
systemd[1]: Stopped target Paths.
systemd[1]: Stopping Paths.
systemd[1]: Stopped target Sockets.
systemd[1]: Stopping Sockets.
systemd[1]: Listening on Journal Audit Socket.
systemd[1]: Stopped target Slices.
systemd[1]: Stopping Slices.
systemd[1]: Started Setup Virtual Console.
systemd[1]: Stopped udev Kernel Device Manager.
systemd[1]: Stopped Create Static Device Nodes in /dev.
systemd[1]: Stopping Create Static Device Nodes in /dev...
systemd[1]: Stopped Create list of required static device nodes for the current kernel.
systemd[1]: Stopping Create list of required static device nodes for the current kernel...
systemd[1]: Closed udev Kernel Socket.
systemd[1]: Stopping udev Kernel Socket.
systemd[1]: Closed udev Control Socket.
systemd[1]: Stopping udev Control Socket.
systemd[1]: Starting Cleanup udevd DB...
systemd[1]: Started Cleanup udevd DB.
systemd[1]: Reached target Switch Root.
systemd[1]: Starting Switch Root.
systemd[1]: Starting Switch Root...
systemd[1]: Switching root.
systemd-journal[76]: Journal stopped
systemd-journal[342]: Runtime journal is using 8.0M (max allowed 395.0M, trying to leave 592.5M free of 3.8G available → current limit 395.0M).
systemd-journal[342]: Permanent journal is using 11.5M (max allowed 4.0G, trying to leave 4.0G free of 203.0G available → current limit 4.0G).
systemd-journal[342]: Time spent on flushing to /var is 16.779ms for 1387 entries.
systemd-journald[76]: Received SIGTERM from PID 1 (systemd).
systemd-journal[342]: Journal started
systemd[1]: systemd 222 running in system mode. (+PAM -AUDIT -SELINUX -IMA -APPARMOR +SMACK -SYSVINIT +UTMP +LIBCRYPTSETUP +GCRYPT +GNUTLS +ACL +XZ +LZ4 +SECCOMP
systemd[1]: Started udev Kernel Device Manager.
systemd[1]: Found device TOSHIBA_DT01ACA300.
systemd[1]: var.mount: Directory /var to mount over is not empty, mounting anyway.
systemd[1]: Mounting /var...
systemd[1]: Mounted /var.
systemd[1]: Mounted Virtual Machine and Container Storage.
systemd[1]: Starting Flush Journal to Persistent Storage...
systemd[1]: Reached target Local File Systems.
systemd[1]: Starting Local File Systems.
systemd[1]: Started Commit a transient machine-id on disk.
systemd[1]: Started Rebuild Journal Catalog.
systemd[1]: Started Update is Completed.
```

**Comment by poettering** [comment]
So, I was confused earlier: what actually happens is that journald won't write to /var at all, ever until it receives SIGUSR1 at least once, which is the indication to use /var if that's configured. This is done by the systemd-journal-flush.service. Hence: there's actually no need to really mount /var read-only early on, because as long as SIGUSR1 is not sent journald won't care about /var...

This also means that Storage=auto is not as borked as I claimed, but I still don't like it...

Not sure what this leaves of this issue report?

Is this now just the ACL race issue that is also discussed in #1397 and #1398? If so, I figure we can close this issue in favour of that one. Doing so now.
